diff --git a/.gitignore b/.gitignore
index 11d0677..ff61dcf 100644
--- a/.gitignore
+++ b/.gitignore
@@ -25,3 +25,4 @@ barrier
 .DS_Store
 *.dSYM
 *.pcap
+.vscode/
\ No newline at end of file
diff --git a/Makefile b/Makefile
index 74e454f..8f6f7dc 100644
--- a/Makefile
+++ b/Makefile
@@ -194,9 +194,11 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
-
-
-
+	$U/_sleep\
+	$U/_pingpong\
+	$U/_primes\
+	$U/_find\
+	$U/_xargs\
 
 ifeq ($(LAB),syscall)
 UPROGS += \
diff --git a/time.txt b/time.txt
new file mode 100644
index 0000000..9a03714
--- /dev/null
+++ b/time.txt
@@ -0,0 +1 @@
+10
\ No newline at end of file
diff --git a/user/find.c b/user/find.c
new file mode 100644
index 0000000..2cbf72d
--- /dev/null
+++ b/user/find.c
@@ -0,0 +1,86 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/fs.h"
+#include "kernel/fcntl.h"
+
+char*
+fmtname(char *path)
+{
+  static char buf[DIRSIZ+1];
+  char *p;
+
+  // Find first character after last slash.
+  for(p=path+strlen(path); p >= path && *p != '/'; p--)
+    ;
+  p++;
+
+  // Return blank-padded name.
+  if(strlen(p) >= DIRSIZ)
+    return p;
+  memmove(buf, p, strlen(p));
+  memset(buf+strlen(p), ' ', DIRSIZ-strlen(p));
+  return buf;
+}
+
+void
+find(char *path, char *name)
+{
+  char buf[512], *p;
+  int fd;
+  struct dirent de;
+  struct stat st;
+
+  if((fd = open(path, O_RDONLY)) < 0){
+    fprintf(2, "find: cannot open %s\n", path);
+    return;
+  }
+
+  if(fstat(fd, &st) < 0){
+    fprintf(2, "find: cannot stat %s\n", path);
+    close(fd);
+    return;
+  }
+
+  if(st.type != T_DIR)
+  {
+    fprintf(2, "find: invalid directory\n");
+    close(fd);
+    return;
+  }
+
+    strcpy(buf, path);
+    p = buf+strlen(buf);
+    *p++ = '/';
+    while(read(fd, &de, sizeof(de)) == sizeof(de)){
+      if(de.inum == 0)
+        continue;
+      memmove(p, de.name, DIRSIZ);
+      p[DIRSIZ] = 0;
+      if(stat(buf, &st) < 0){
+        printf("find: cannot stat %s\n", buf);
+        continue;
+      }
+
+    if(st.type==T_DIR && strcmp(p, ".")!=0 && strcmp(p, "..")!=0)
+    {
+        find(buf, name);
+    } else if(st.type==T_FILE && strcmp(p, name)==0)
+    {
+        printf("%s\n", buf);
+    }
+  }
+  close(fd);
+}
+
+int 
+main(int argc, char *argv[])
+{
+  if(argc != 3)
+  {
+    fprintf(2, "Invalid usage!\n");
+    exit(1);
+  }
+  find(argv[1], argv[2]);
+  exit(0);
+}
diff --git a/user/pingpong.c b/user/pingpong.c
new file mode 100644
index 0000000..92f4c21
--- /dev/null
+++ b/user/pingpong.c
@@ -0,0 +1,57 @@
+#include "kernel/types.h"
+#include "user.h"
+
+int main(int argc, char* argv[]) {
+    int pipePing[2], pipePong[2];
+    if(pipe(pipePing) == -1 || pipe(pipePong) == -1) {
+        fprintf(2, "Error: can't create pipe\n");
+        exit(1);
+    }
+
+
+    char buffer = 'c';
+    int pid = fork();
+    if(pid < 0) {
+        fprintf(2, "Error: can't fork\n");
+        exit(1);
+    }
+
+    if(pid == 0) { // child process
+        close(pipePing[1]);
+        close(pipePong[0]);
+
+        if(read(pipePing[0], &buffer, 1) == -1) {
+            fprintf(2, "Error: child can't read\n");
+            exit(1);
+        }
+        fprintf(1, "%d: received ping\n", getpid());
+        close(pipePing[0]);
+
+        if(write(pipePong[1], &buffer, 1) == -1) {
+            fprintf(2, "Error: child can't write\n");
+            exit(1);
+        }
+        close(pipePong[1]);
+        exit(0);
+    }
+
+    else { // parent process
+        close(pipePing[0]);
+        close(pipePong[1]);
+        
+        if(write(pipePing[1], &buffer, 1) == -1) {
+            fprintf(2, "Error: parent can't write\n");
+            exit(1);
+        }
+        close(pipePing[1]);
+
+        if(read(pipePong[0], &buffer, 1) == -1) {
+            fprintf(2, "Error: parent can't read\n");
+            exit(1);
+        }
+        fprintf(1, "%d: received pong\n", getpid());
+        close(pipePong[0]);
+        wait(0);
+        exit(0);
+    }
+}
\ No newline at end of file
diff --git a/user/primes.c b/user/primes.c
new file mode 100644
index 0000000..ba8332e
--- /dev/null
+++ b/user/primes.c
@@ -0,0 +1,54 @@
+#include "kernel/types.h"
+#include "user.h"
+
+void sieve(int pipeLeft[2]) __attribute__((noreturn));
+
+void sieve(int pipeLeft[2]) {
+    close(pipeLeft[1]);
+    int prime, n;
+    
+    if(read(pipeLeft[0], &prime, sizeof(int)) == 0) {
+        exit(0);
+    }
+    fprintf(1, "prime %d\n", prime);
+    
+    int pipeRight[2];
+    pipe(pipeRight);
+    int pid = fork();
+    if(pid) {
+        close(pipeRight[0]);
+        while(read(pipeLeft[0], &n, sizeof(int)) > 0) {
+            if(n % prime != 0) {
+                write(pipeRight[1], &n, sizeof(int));
+            }
+        }
+        close(pipeRight[1]);
+        close(pipeLeft[0]);
+        wait(0);
+        exit(0);
+    } else {
+        close(pipeLeft[0]);
+        close(pipeRight[1]);
+        sieve(pipeRight);
+    }
+}
+
+int main(int argc, char* argv[]) {
+    int pipeMain[2];
+    pipe(pipeMain);
+
+    int pid = fork();
+    if(!pid) {
+        close(pipeMain[1]);
+        sieve(pipeMain);
+    } else {
+        close(pipeMain[0]);
+        for(int i = 2; i <= 280; i++) {
+            write(pipeMain[1], &i, sizeof(int));
+        }
+        close(pipeMain[1]);
+        wait(0); 
+    }
+    
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/sleep.c b/user/sleep.c
new file mode 100644
index 0000000..3fa4a64
--- /dev/null
+++ b/user/sleep.c
@@ -0,0 +1,13 @@
+#include "kernel/types.h"
+#include "user.h"
+
+int main(int argc, char* argv[]) {
+    if(argc != 2) {
+        fprintf(2, "Usage: sleep <ticks>\n");
+        exit(1);
+    }
+
+    int n = atoi(argv[1]);
+    sleep(n);
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/xargs.c b/user/xargs.c
new file mode 100644
index 0000000..624b854
--- /dev/null
+++ b/user/xargs.c
@@ -0,0 +1,69 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/param.h"
+
+#define MAXLINE 512 
+
+int getcmd(char *buf, int nbuf) {
+    memset(buf, 0, nbuf);
+    int n = read(0, buf, nbuf - 1);
+    if (n <= 0) return -1;
+    buf[n] = 0;
+    return 0;
+}
+
+char whitespace[] = " \t\r\n\v";
+
+int gettoken(char **ps, char *es, char **q, char **eq) {
+    char *s = *ps;
+
+    while (s < es && strchr(whitespace, *s)) s++;
+    if (q) *q = s;
+
+    int ret = *s;
+    if (*s) {  
+        ret = 'a';
+        while (s < es && !strchr(whitespace, *s)) s++;
+    }
+
+    if (eq) *eq = s;
+    if (s < es) *s++ = 0;
+
+    while (s < es && strchr(whitespace, *s)) s++;
+    *ps = s;
+    return ret;
+}
+
+int main(int argc, char *argv[]) {
+    char *xargs[MAXARG]; 
+    for (int i = 1; i < argc; i++) {
+        xargs[i - 1] = argv[i];  
+    }
+
+    static char buf[MAXLINE];
+    char *q, *eq;
+
+    while (getcmd(buf, sizeof(buf)) >= 0) {
+        int j = argc - 1;  
+        char *s = buf;
+        char *es = s + strlen(s);
+
+        while (gettoken(&s, es, &q, &eq) != 0) {
+            *eq = 0;  
+            xargs[j] = q;
+            xargs[j + 1] = 0;  
+
+            int pid = fork();
+            if (pid == 0) {  
+                exec(xargs[0], xargs);
+                fprintf(2, "xargs: exec failed for %s\n", xargs[0]);
+                exit(1);
+            } else {  
+                wait(0);
+            }
+        }
+    }
+
+    exit(0);
+}
